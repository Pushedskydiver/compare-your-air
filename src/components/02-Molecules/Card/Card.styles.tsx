import styled from 'styled-components';
import { colors } from '../../../styles/colors.styles';
import spacing from '../../../styles/spacing.styles';

const CardStyles = styled.li`
  position: relative;
  padding: ${spacing.padding * 1.75}px;
  border-radius: 8px;
  background-color: ${colors.light};
  color: ${colors.dark};
`;


export default CardStyles;
