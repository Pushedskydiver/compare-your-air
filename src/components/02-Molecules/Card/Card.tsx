import React, { useContext } from 'react';
import CardStyles from './Card.styles';
import Button from '../../01-Atoms/Button/Button.styles';
import { Close } from '../../01-Atoms/Icon/Icon';
import { Heading2 } from '../../01-Atoms/Heading/Heading.styles';
import Paragraph from '../../01-Atoms/Paragraph/Paragraph.styles'
import { colors } from '../../../styles/colors.styles';
import { calcLastUpdated, formatDate } from '../../Utils/Utils';
import Context from '../../Context';

type PropsForCard = {
  location: string;
  city: string;
  measurements: Array<any>;
};

type PropsForValues = {
  parameter: string;
  value: string;
};

type PropsForLastUpdated = {
  lastUpdated: string;
}

function renderValues(measurement: PropsForValues) {
  const parameter = measurement.parameter;
  const value = measurement.value;

  return `${parameter}: ${value}`;
}

function renderLastUpdated(measurement: PropsForLastUpdated) {
  const todaysDate = new Date();
  const previousDate = new Date(measurement.lastUpdated);
  const lastUpdated: any = calcLastUpdated(todaysDate, previousDate);
  const filteredDate = Object.keys(lastUpdated)
    .filter(key => lastUpdated[key] > 0)
    .map(key => formatDate(key, lastUpdated[key]));
  
  return filteredDate[0];
}

const Card = (props: PropsForCard) => {
  const { removeSelectedLocation } = useContext(Context);
  const { location, city, measurements } = props;
  const values = measurements.map(renderValues).join(', ');
  const lastUpdated = renderLastUpdated(measurements[0]);

  function handleCardRemoval(event: any) {
    const location = event.target.getAttribute('data-location');

    removeSelectedLocation(location);
  }
  
  return (
    <CardStyles>
      <Button
        aria-label={`Remove ${location} from the selected locations list.`}
        data-location={location}
        onClick={handleCardRemoval}
      >
        <Close/>
      </Button>

      <Paragraph bold noMargin small uppercase>Updated {lastUpdated}</Paragraph>
      <Heading2 color={colors.core.alpha} medium>{location}</Heading2>

      <Paragraph medium>In {city}, United Kingdom</Paragraph>
      <Paragraph bold uppercase><strong>Values: {values}</strong></Paragraph>
    </CardStyles>
  );
}

export default Card;
