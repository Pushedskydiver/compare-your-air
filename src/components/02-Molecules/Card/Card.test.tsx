import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Card from './Card';

const measurements = [
  {
    parameter: 'no2',
    value: 27,
    lastUpdated: '2020-02-15T10:00:00.000Z',
    unit: 'µg/m³',
    sourceName: 'DEFRA',
    averagingPeriod: {
      value: 1,
      unit: 'hours'
    }
  },
]

afterEach(() => {
  cleanup();
});

describe('<Card />', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <Card location="Manchester Piccadilly" city="Manchester" measurements={measurements} />
    );

    expect(container).toMatchSnapshot();
  });
});
