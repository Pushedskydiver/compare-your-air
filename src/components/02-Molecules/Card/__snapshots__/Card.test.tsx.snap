// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should match snapshot 1`] = `
<div>
  <li
    class="sc-AxhCb gBdabp"
    data-testid="location-card"
  >
    <button
      aria-label="Remove Manchester Piccadilly from the selected locations list."
      class="sc-AxhUy vrqkV"
      data-location="Manchester Piccadilly"
      data-testid="location-remove"
    >
      <svg
        class="sc-AxgMl ddozpw"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.984 6.422L13.406 12l5.578 5.578-1.406 1.406L12 13.406l-5.578 5.578-1.406-1.406L10.594 12 5.016 6.422l1.406-1.406L12 10.594l5.578-5.578z"
        />
      </svg>
    </button>
    <p
      class="sc-fzpans iqBrKi"
    >
      Updated 
      2 days ago
    </p>
    <h2
      class="sc-Axmtr bZYwdh"
      color="#7430a2"
    >
      Manchester Piccadilly
    </h2>
    <p
      class="sc-fzpans jgOCBH"
    >
      In 
      Manchester
      , United Kingdom
    </p>
    <p
      class="sc-fzpans hZXKAf"
    >
      <strong>
        Values: 
        no2: 27
      </strong>
    </p>
  </li>
</div>
`;
